source/class_binder.c

```diff
@@ -0,0 +1,431 @@
+        case napi_boolean: {
+            AWS_NAPI_CALL(env, napi_get_value_bool(env, value, &out_value->native.boolean), { return status; });
+
+            break;
+        }
+
+        case napi_string: {
+            AWS_NAPI_CALL(env, aws_byte_buf_init_from_napi(&out_value->native.string, env, value), { return status; });
+
+            break;
+        }
+
+        case napi_number: {
+            AWS_NAPI_CALL(env, napi_get_value_int64(env, value, &out_value->native.number), {
+                napi_throw_type_error(env, NULL, "Class binder argument expected a number");
+                return status;
+            });
+
+            break;
+        }
+
+        case napi_external: {
+            AWS_NAPI_CALL(env, napi_get_value_external(env, value, &out_value->native.external), {
+                napi_throw_type_error(env, NULL, "Class binder argument expected an external");
+                return status;
+            });
+
+            break;
+        }
+
+        case napi_object: {
+            /* Attempt to unwrap the object, just in case */
+            napi_status result = napi_unwrap(env, value, &out_value->native.external);
+            if (result != napi_ok) {
+                out_value->native.external = NULL;
+            }
+
+            break;
+        }
+
+        default:
+            /* Don't process, just leave as node value */
+            break;
+    }
+
+    return napi_ok;
+}
+
+/**
+ * Cleans up an aws_napi_argument object populated by s_argument_parse.
+ *
+ * \param env           The node environment.
+ * \param value         The value to clean up.
+ */
+static void s_argument_cleanup(napi_env env, struct aws_napi_argument *value) {
+    (void)env;
+
+    switch (value->type) {
+        case napi_string:
+            aws_byte_buf_clean_up(&value->native.string);
+            break;
+
+        default:
+            break;
+    }
+}
+
+/**
+ * Used as the class's constructor
+ */
+static napi_value s_constructor(napi_env env, napi_callback_info info) {
+
+    napi_value node_args[AWS_NAPI_METHOD_MAX_ARGS];
+    napi_value node_this = NULL;
+    size_t num_args = AWS_ARRAY_SIZE(node_args);
+    struct aws_napi_class_info_impl *clazz = NULL;
+    AWS_NAPI_CALL(env, napi_get_cb_info(env, info, &num_args, node_args, &node_this, (void **)&clazz), {
+        napi_throw_error(env, NULL, "Failed to retreive callback information");
+        return NULL;
+    });
+    if (num_args > AWS_NAPI_METHOD_MAX_ARGS) {
+        num_args = AWS_NAPI_METHOD_MAX_ARGS;
+    }
+
+    napi_value result = NULL;
+
+    /* Check if we're wrapping an existing object or creating a new one */
+    if (clazz->is_wrapping) {
+        AWS_FATAL_ASSERT(num_args == 1);
+
+        void *native = NULL;
+
+        /* Arg 1 should be an external */
+        AWS_NAPI_ENSURE(env, napi_get_value_external(env, node_args[0], &native));
+
+        /* Wrap shouldn't take a finalizer, because it's very likely that this object isn't owned by JS */
```

Comment:
Note that the external has the finalizer

Reply:
OOPS this is phat wrong